service-base/src/main/java/org/eclipse/hono/service/command/CommandConnectionImpl.java

```diff
@@ -0,0 +1,105 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.service.command;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import org.eclipse.hono.client.impl.HonoClientImpl;
+import org.eclipse.hono.config.ClientConfigProperties;
+import org.eclipse.hono.connection.ConnectionFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.proton.ProtonDelivery;
+
+/**
+ * Implements a connection between an Adapter and the AMQP 1.0 network to receive commands and send a response.
+ */
+public class CommandConnectionImpl extends HonoClientImpl implements CommandConnection {
+
+    protected final Logger LOG = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Creates a new client for a set of configuration properties.
+     * <p>
+     * This constructor creates a connection factory using
+     * {@link ConnectionFactory#newConnectionFactory(Vertx, ClientConfigProperties)}.
+     *
+     * @param vertx The Vert.x instance to execute the client on, if {@code null} a new Vert.x instance is used.
+     * @param clientConfigProperties The configuration properties to use.
+     * @throws NullPointerException if clientConfigProperties is {@code null}
+     */
+    public CommandConnectionImpl(final Vertx vertx, final ClientConfigProperties clientConfigProperties) {
+        super(vertx, clientConfigProperties);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public Future<CommandAdapter> createCommandConsumer(
+            final String tenantId,
+            final String deviceId,
+            final Consumer<Command> messageConsumer,
+            final Handler<Void> closeHandler) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(messageConsumer);
+        return checkConnected().compose(con -> {
+            final Future<CommandAdapter> result = Future.future();
+            CommandAdapter.create(context, clientConfigProperties, connection, tenantId, deviceId,
```

Comment:
Possibly the `create` method would be better located in the `Command` class?
In the end, the creation of the adapter results here in the subsequent creation of the `Command` anyway.

Maybe I miss something, but the dependencies between `Command`, `CommandAdapter` and `CommandConnection` seem to me more complicated as they should - WDYT?

Reply:
The CommandConnection does the connection handling. On that connection are n link pairs, which are included in the CommandAdapter. On this receiver links there could arrive n commands, which need to send back a response on the sender link (included in the CommandAdapter).

What do you suggest to simplification?

Reply:
After having a second look at the code, let me first ask the question:

why would a protocol adapter ever deal with a `CommandAdapter` directly?
IMHO it handles a received command (instance of `Command`), and it may handle a response to it (by passing the instance of `Command` to the `sendCommandResponse` method of `CommandConnection`). 
So it works with `Command` and `CommandConnection`. 

Is that right?

Reply:
The needed way to close a receiver (and the sender) is now available with a MessageConsumer.